typeform/typeform.py

```diff
@@ -18,188 +30,163 @@ class Typeform(panoply.DataSource):
-        # allow incremental pulling from the last successful batch
-        incval = source.get('lastTimeSucceed')
-        self._incval = getTimestamp(incval)
+        source['destination'] = source.get('destination') or DESTINATION
+        # append the destination postfix
+        if DESTINATION_POSTFIX not in source['destination']:
+            source['destination'] += '_{}'.format(DESTINATION_POSTFIX)
 
+        self._incval = get_incval(source)
 
         forms = source.get('forms', [])
         self._forms = copy.deepcopy(forms)
 
-        # add an 'offset' attribute used for pagination for each
-        # different form, used as the actual number to use as offset
-        for f in self._forms:
-            f['offset'] = 0
+        # add an 'after' attribute used
+        # for pagination for each different form
+        for form in self._forms:
+            form['after'] = None
 
-        self._key = source.get('key')
+        self._access_token = source.get('access_token')
         self._total = len(self._forms)
 
-    def read(self, n = None):
-        if len(self._forms) == 0:
+    def read(self, n=None):
+        if not self._forms:
             # no more data to consume
             return None
 
         form = self._forms[0]
 
         # construct the GET url and make the request
-        url = self._url(form)
-        body = self._request(url)
-
-        # 'completed' represent the number of completed forms that
-        # returned from the server
-        stats = body.get('stats', {}).get('responses', {})
-        total = stats.get('completed', 0)
+        params = self._build_params(form, n)
+        url = FORM_RESPONSES_URL.format(**form)
+        response = self._request(url, params)
 
-        if (form['offset'] + FETCH_LIMIT) >= total:
+        items = response.get('items', [])
+        if not items:
             # we're done paginating with the current form.
-            form['offset'] = None
-
             # no more results for this form, remove it
             self._forms.pop(0)
-
-            # report progress
-            loaded = self._total - len(self._forms)
-            msg = '%s of %s forms fetched' % (loaded, self._total)
-            self.progress(loaded, self._total, msg)
         else:
             # prepare the offset to the next set of records
-            form['offset'] += FETCH_LIMIT
-
-
-        # When fetching data from the Typeform API, it provides us with
-        # the form questions, responses and general statistics. Instead
-        # of creating multiple tables for each form, We want to create
-        # a single set of tables for the aforementioned statistics.
-        # Note that each row is associated to it's form name,
-        # indicated by the '__form' column.
-
-        # generate a single result statistics record from
-        # the recent pulling
-        stats = {
-            '__table': 'stats',
-            '__form': form.get('name'),
-            'total': stats.get('total'),
```

Comment:
I don't see 'total' in the updated code. Did we plan to remove it?

Reply:
Theres isn't 'total` attribute anymore.
There is `total_items` attribute (is the number of responses your request retrieved)
When we paginate `total_items` is changing on each response